### PR TITLE
Augur Mask: Add additional options from NCOV mask-alignment.py script

### PR DESCRIPTION
**Pull request by danielsoneg**
### Description of proposed changes    
This PR adds the `--mask-sites`, `--mask-from-beginning`, and `--mask-from-end` args to `augur mask`.

`--mask-from-beginning` and `--mask-from-end` are gated to only apply to FASTA files.

### Related issue(s)  

Related to conversation in #459 and #148 

### Testing
Ran full test suite, updated tests to test for new cases.

Also expanded existing test coverage, now at 100% coverage 🕺

### CODE REVIEW

**Comment by huddlej** [comment]
On first pass, this looks great! I'll test it out more tonight and post some feedback.

File: augur/mask.py [hunk 1]
```diff
@@ -122,7 +134,10 @@ def mask_fasta(mask_sites, in_file, out_file):
 
 def register_arguments(parser):
     parser.add_argument('--sequences', '-s', required=True, help="sequences in VCF or FASTA format")
-    parser.add_argument('--mask', required=True, help="locations to be masked in BED file format")
+    parser.add_argument('--mask', dest="mask_file", required=False, help="locations to be masked in BED file format")
+    parser.add_argument('--mask-from-beginning', type=int, help="FASTA Only: Number of sites to mask from beginning")
+    parser.add_argument('--mask-from-end', type=int, help="FASTA Only: Number of sites to mask from end")
```
Comment (huddlej) [inline augur/mask.py:139]: These two arguments need to have default values of 0 or they will get passed as `None` to the downstream math in `mask_fasta` and produce this exception:

```
Removing masked sites from FASTA file.
Traceback (most recent call last):
  File "./bin/augur", line 24, in <module>
    exit( main() )
  File "./augur/__main__.py", line 10, in main
    return augur.run( argv[1:] )
  File "./augur/__init__.py", line 74, in run
    return args.__command__.run(args)
  File "./augur/mask.py", line 197, in run
    mask_from_end=args.mask_from_end)
  File "./augur/mask.py", line 121, in mask_fasta
    if beginning + end > sequence_length:
TypeError: unsupported operand type(s) for +: 'NoneType' and 'NoneType'
```

File: tests/test_mask.py [hunk 2]
```diff
@@ -225,15 +284,46 @@ def make_outfile(mask_sites, in_file, out_file, cleanup=True):
...
+        args = argparser("--mask=%s --sequences=%s --output=%s" %(bed_file, vcf_file, out_file))
+        mask.run(args)
+        assert os.path.exists(out_file), "Output file incorrectly deleted"
+    
+    def test_run_with_mask_sites(self, vcf_file, out_file, argparser, mp_context):
+        args = argparser("--mask-sites 2 8 -s %s -o %s" % (vcf_file, out_file))
+        def check_mask_sites(mask_sites, *args, **kwargs):
+            assert mask_sites == [2,8]
+        mp_context.setattr(mask, "mask_vcf", check_mask_sites)
+        mask.run(args)
```
Comment (huddlej) [inline tests/test_mask.py:303]: This test would be more helpful if it ran `mask_vcf` and confirmed that the requested sites were no longer present in the output file. I actually love the way you've setup the argparser fixture here such that we can essentially run end-to-end tests from the unit testing framework. I know it is a little more computationally expensive, but running the whole program this way can catch more errors.

We also need a corresponding test for masking specific sites with a FASTA input. This test will fail right now because mask_sites values are not converted to 0-based coordinates (and because of the mask from beginning/end None issue mentioned earlier).

```
def test_run_with_mask_sites_for_fasta(self, fasta_file, out_file, argparser):
    # 1-based coordinates from the user should translate to 0-based masking of the FASTA sequence.
    args = argparser("--mask-sites 1 -s %s -o %s" % (fasta_file, out_file))
    mask.run(args)
    output = SeqIO.parse(out_file, "fasta")
    for record in output:
        # Site 1 in the sequence is the 0-indexed site in the sequence record.
        assert record.seq[0] == "N"
```

**Comment by danielsoneg** [comment]
Actually, let me rephrase this to be a bit clearer. Let's say we have the sequence below:

```
CCCCCCCCC # Sequence
012345678 # Zero-based indexes
```

So, here's my understanding of the state of the world:
### 1. Sites passed into the CLI
if someone issues: `augur mask --mask-sites 1 3`, the expected output is:
```
NCNCCCCCC # Sequence
012345678 # Zero-based indexes
```

### 2. Sites passed in via a masking file:
Someone issues `augur mask --mask test.not_a_bed_file` (I'm adding that next, so)
test.not_a_bed_file contains:
```
4
5
```
Our expected output is:
```
CCCNNCCCC # Sequence
012345678 # Zero-based indexes
```
### 3. Sites passed in via a BED file
This is where I start to get confused - let's say a user passes in `augur mask --mask test.bed`.
test.bed contains:
```
SEQ	5	7
```
The current code will output this for **FASTA**:
```
CCCCCNNNC # Sequence
012345678 # Zero-based indexes
```
And this for **VCF**:
```
CCCCNNNCC # Sequence
012345678 # Zero-based indexes
```
So it's obviously doing the wrong thing for one of these, but what's the correct output?

(I'm noting I missed this because indeed any line that refers to the specified index is gone from the VCF, that just means something different for VCFs)

File: tests/test_mask.py (hunk 2 above)
Comment (huddlej) [reply tests/test_mask.py]: That is so cool! Thank you!

**Comment by huddlej** [comment]
The rephrasing in your last comment is super helpful! Parts 1 and 2 are exactly right. Part 3 is complicated by the bug in [how BED files are currently read into augur](https://github.com/nextstrain/augur/issues/511). So there is what the output should be for the current augur code and then there is what the output should be for a properly implemented BED parser. I'll list both outputs below.

If the input BED file looks like this:

```
SEQ	5	7
```

The FASTA output for the current augur code should be:

```
CCCCCNNNC # Sequence
012345678 # Zero-based indexes
```

The VCF output for the current augur code should be:

```
CCCCCNNNC # Sequence
012345678 # Zero-based indexes
123456789 # One-based indexes ([6, 7, 8] would be written out for vcftools)
```

The FASTA output for a proper 0-indexed, half-open BED parser should be:

```
CCCCCNNCC # Sequence
012345678 # Zero-based indexes
```

The last position in the interval isn't masked because BED is half-open. Resolving that bug can be a future PR though.

Thanks for also adding the end-to-end tests! I'm going to use these to start a proposal for how we more generally implement end-to-end tests in augur along with other possible approaches like the [cram tool](https://bitheap.org/cram/) @tsibley has recommended.

**Comment by danielsoneg** [comment]
Ok, that's perfect - thanks! 

I can fix the BED bug in this pr - it's a pretty quick fix. I'll also fix this in #514 as well.

**Comment by huddlej** [comment]
@danielsoneg Can you hold off from fixing the BED parsing bug? I think everything you've updated in augur mask so far is still backwards-compatible, but fixing the BED parsing issue will change the data API.

I want to make sure we coordinate with @emmahodcroft when she has a free moment, so we don't break her builds that rely on the current way BED files are parsed.

**Comment by danielsoneg** [comment]
Ok, updated the diff - `mask_vcf` is now taking the zero-indexed `mask_site` lists and incrementing them all by 1 before writing out the masking file that's passed to vcftools.

For what it's worth, I'm 95% sure this is a bug that's existed in mask.py's BED file handling for a while now - tree.py's been fine because treetime.read_vcf already zero-indexes bed files, but I think even adjusting the indexing for the mask file is a change in behavior.

**Comment by huddlej** [comment]
Ah, darn, you are totally right. Converting the masks to 1-indexed sites for vcftools is already a change in API, so we might as well fix the half-open bug for BED files here, too. Thanks for catching that!

This means I'll need to work through one of @emmahodcroft's affected builds and verify the output with these new modifications before we merge this PR.

**Comment by danielsoneg** [comment]
Ok, updated with the BED fixes as well

**Comment by huddlej** [comment]
Perfect! I will work on testing these tomorrow afternoon.

**Comment by huddlej** [comment]
Sorry for the super delayed response here, @danielsoneg, but this worked well for me locally. I also dropped this new augur mask functionality into a local ncov build to replace the custom script it is currently using and I got exactly the same results.

Merging this now and I'll look into the tree PR next. Thanks again for your help here!

**Comment by danielsoneg** [comment]
No prob - let me take a look at the tree one before you review, though. It's been a while and I don't think that had the BED file changes - I'll ping you once I'm sure it's in good shape
